### PR TITLE
Servers have to change connection IDs in Retry

### LINKED ISSUE #2837: Does a Retry really need to change the CID?
From section 17.2.5: https://tools.ietf.org/html/draft-ietf-quic-transport-20#section-17.2.5
"This value MUST not be equal to the Destination Connection ID field of the packet sent by the client."

This was originally added when multiple Retrys were supported.  I don't think it's still necessary, but if it is, I'd like to add some rationale, because it's not obvious to me why this restriction is necessary.

Past slides from Martin: https://github.com/quicwg/wg-materials/blob/master/interim-18-09/retry.pdf

### PR DESCRIPTION
**Pull request by martinthomson**
Two changes here: better explanation of why, and a validation requirement on clients.

Closes #2837.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1636,10 +1636,14 @@ As long as it is not possible for an attacker to generate a valid token for
...
-use, a server can cause the connection to be routed to a server instance with
-more resources available for new connections.
+A server can also use a Retry packet to defer the state and processing costs of
+connection establishment.  Requiring the server to provide a different
+connection ID, along with the original_connection_id transport parameter defined
+in {{transport-parameter-definitions}}, forces the server to demonstrate that
+it, or an entity it cooperates with, received the original Initial packet from
+the client.  Providing a different connection ID also grants a server some
+control over how subsequent packets are routed.  This can be used to direct
+connections to a server instance with more resources available.
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]:
```suggestion
connections to a different server instance.
```

Comment (ianswett) [reply draft-ietf-quic-transport.md]: There are lots of possible reasons why, so I wouldn't attempt to enumerate them.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -4150,8 +4154,10 @@ the client included in the Source Connection ID of the Initial packet.
 
 The server includes a connection ID of its choice in the Source Connection ID
 field.  This value MUST not be equal to the Destination Connection ID field of
-the packet sent by the client.  The client MUST use this connection ID in the
-Destination Connection ID of subsequent packets that it sends.
+the packet sent by the client.  A client MUST discard a Retry packet that
+contains a Source Connection ID field that is identical to the Destination
+Connection ID field of its Initial packet.  The client MUST use this connection
+ID in the Destination Connection ID field of subsequent packets that it sends.
```
Comment (mikkelfj) [inline draft-ietf-quic-transport.md]: Post merge, but "The client MIST use this connection ID" reads as if it refers to the invalid duplicate ID.

Comment (martinthomson) [reply draft-ietf-quic-transport.md]: https://github.com/quicwg/base-drafts/pull/3153

**Pull request merged**